src/pybind/mgr/osd_support/module.py

```diff
@@ -0,0 +1,330 @@
+To stop and reset the weight of already started operations we need to save the initial weight 
+(see 'Ideas for improvement')
+
+
+Ideas for improvement:
+- add health checks set_health_checks
+- use objects to represent OSDs
+  - allows timestamps, trending information etc
+- save osd drain state (at least the osd_ids in the mon store)
+  - resume after a mgr crash
+  - save the initial weight of a osd i.e. (set to initial weight on abort)
+"""
+
+
+class OSDSupport(MgrModule):
+    # these are CLI commands we implement
+    COMMANDS = [
+        {
+            "cmd": "osd drain name=osd_ids,type=CephInt,req=true,n=N",
+            "desc": "drain osd ids",
+            "perm": "r"
+        },
+        {
+            "cmd": "osd drain status",
+            "desc": "show status",
+            "perm": "r"
+        },
+        {
+            "cmd": "osd drain stop name=osd_ids,type=CephInt,req=false,n=N",
+            "desc": "show status for osds. Stopping all if osd_ids are omitted",
+            "perm": "r"
+        },
+    ]
+
+    # These are module options we understand.  These can be set with
+    #
+    #   ceph config set global mgr/hello/<name> <value>
+    #
+    # e.g.,
+    #
+    #   ceph config set global mgr/hello/place Earth
+    #
+    MODULE_OPTIONS: List[dict] = []
+
+    # These are "native" Ceph options that this module cares about.
+    NATIVE_OPTIONS: List[str] = [
+        'mgr_tick_period',
+    ]
+
+    osd_ids: Set[int] = set()
+    emptying_osds: Set[int] = set()
+    check_osds: Set[int] = set()
+
+    def __init__(self, *args, **kwargs):
+        super(OSDSupport, self).__init__(*args, **kwargs)
+
+        # set up some members to enable the serve() method and shutdown()
+        self.run = True
+        self.event = Event()
+
+        # ensure config options members are initialized; see config_notify()
+        self.config_notify()
+
+    def config_notify(self):
+        """
+        This method is called whenever one of our config options is changed.
+        """
+        # This is some boilerplate that stores MODULE_OPTIONS in a class
+        # member, so that, for instance, the 'emphatic' option is always
+        # available as 'self.emphatic'.
+        for opt in self.MODULE_OPTIONS:
+            setattr(self,
+                    opt['name'],
+                    self.get_module_option(opt['name']) or opt['default'])
+            self.log.debug(' mgr option %s = %s',
+                           opt['name'], getattr(self, opt['name']))
+        # Do the same for the native options.
+        for opt in self.NATIVE_OPTIONS:
+            setattr(self,
+                    opt,
+                    self.get_ceph_option(opt))
+            self.log.debug(' native option %s = %s', opt, getattr(self, opt))
+
+    def handle_command(self, inbuf, cmd):
+        ret = 0
+        err = ''
+        _ = inbuf
+        cmd_prefix = cmd.get('prefix', '')
+        osd_ids: List[int] = cmd.get('osd_ids', list())
```

Comment:
just a hint, this new-style decorator approach for defining commands automatically does this for you.

Reply:
I feel at this point we have to move forward.. I'll consider moving to the new style in a future pr